flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTaskNetworkInput.java

```diff
@@ -197,6 +205,29 @@ public int getInputIndex() {
 		return checkpointedInputGate.getAvailableFuture();
 	}
 
+	@Override
+	public CompletableFuture<?> prepareSnapshot(long checkpointId) throws IOException {
+		// Note that if considering recovery in future, we should guarantee that the spilled buffers in one channel
+		// should be close together because one record might span multiple buffers.
+		for (int channelIndex = 0; channelIndex < recordDeserializers.length; channelIndex++) {
+			final InputChannel channel = checkpointedInputGate.getChannel(channelIndex);
+
+			recordDeserializers[channelIndex].getUnconsumedBuffer().ifPresent(buffer ->
+				channelStateWriter.addInputData(
+					checkpointId,
+					channel.getChannelInfo(),
+					ChannelStateWriter.SEQUENCE_NUMBER_UNKNOWN,
+					buffer));
+
+			channelStateWriter.addInputData(
+				checkpointId,
+				channel.getChannelInfo(),
+				ChannelStateWriter.SEQUENCE_NUMBER_UNKNOWN,
+				checkpointedInputGate.getInflightBuffers(checkpointId, channelIndex).toArray(new Buffer[0]));
+		}
+		return checkpointedInputGate.getAllBarriersReceivedFuture(checkpointId);
+	}
+
 	@Override
 	public void close() throws IOException {
```

Comment:
Probably, we should add to `close` 
`CheckpointedInputGate.close` -> `CheckpointBarrierHandler.close` -> `future.complete`
(and similar places)

Reply:
Interesting idea. However, would trigger `complete` not more harm than it would solve? When we trigger `complete`, it might create an incomplete checkpoint...

Reply:
You're right. What about `completeExceptionally`?